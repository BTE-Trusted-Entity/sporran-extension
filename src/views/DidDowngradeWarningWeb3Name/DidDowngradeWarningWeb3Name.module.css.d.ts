export const __esModule: true;
export const container: string;
export const heading: string;
export const subline: string;
export const warning: string;
export const details: string;
export const label: string;
export const explanation: string;
export const buttonsLine: string;
export const cancel: string;
export const cta: string;
