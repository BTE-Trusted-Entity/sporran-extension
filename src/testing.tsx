export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
