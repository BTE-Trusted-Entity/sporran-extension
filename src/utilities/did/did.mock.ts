import { isFullDid } from './did';

jest.mock('./did');

export function mockIsFullDid(boolean: boolean): void {
  jest.mocked(isFullDid).mockReturnValue(boolean);
}
