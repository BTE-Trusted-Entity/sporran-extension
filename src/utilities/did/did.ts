import {
  ConfigService,
  Did,
  DidDocument,
  DidEncryptionKey,
  DidUri,
  Utils,
} from '@kiltprotocol/sdk-js';

import { useAsyncValue } from '../useAsyncValue/useAsyncValue';

export function isFullDid(did: DidUri | undefined): boolean {
  if (!did) {
    // could be a legacy identity without DID
    return false;
  }
  return Did.parse(did).type === 'full';
}

export async function getFullDidDocument(did: DidUri): Promise<DidDocument> {
  const api = ConfigService.get('api');
  return Did.linkedInfoFromChain(await api.call.did.query(Did.toChain(did)))
    .document;
}

export function useFullDidDocument(did: DidUri): DidDocument | undefined {
  return useAsyncValue(getFullDidDocument, [did]);
}

export async function getDidDocument(did: DidUri): Promise<DidDocument> {
  return isFullDid(did)
    ? await getFullDidDocument(did)
    : Did.parseDocumentFromLightDid(did);
}

export function getDidEncryptionKey({
  keyAgreement,
}: DidDocument): DidEncryptionKey {
  if (!keyAgreement || !keyAgreement[0]) {
    throw new Error('encryptionKey is not defined somehow');
  }
  return keyAgreement[0];
}

export function parseDidUri(did: DidUri): ReturnType<typeof Did.parse> & {
  fullDid: DidUri;
} {
  const parsed = Did.parse(did);
  const fullDid = parsed.type === 'full' ? did : Did.getFullDidUri(did);

  return {
    ...parsed,
    fullDid,
  };
}

export async function needLegacyDidCrypto(
  did: DidUri | undefined,
): Promise<boolean> {
  if (!did) {
    // DID was deactivated, no action needed.
    return false;
  }
  if (!isFullDid(did)) {
    return false;
  }

  try {
    const encryptionKey = getDidEncryptionKey(await getDidDocument(did));
    return (
      Utils.Crypto.u8aToHex(encryptionKey.publicKey) ===
      '0xf2c90875e0630bd1700412341e5e9339a57d2fefdbba08de1cac8db5b4145f6e'
    );
  } catch {
    // getDidDocument might throw if the DID is not on-chain anymore (removed, another endpoint),
    // no legacy crypto needed in that case
    return false;
  }
}
