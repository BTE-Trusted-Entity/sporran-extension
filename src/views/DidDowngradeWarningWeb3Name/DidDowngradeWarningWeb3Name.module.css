.container {
  composes: container from '../../components/View/View.module.css';
}

.heading {
  composes: screenHeadline from '../../components/Typography/Typography.module.css';
}

.subline {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin: 0 0 0.8125rem;
}

.warning {
  composes: importantHeadline from '../../components/Typography/Typography.module.css';
  font-size: 1.25rem;
  margin: 0.75rem 0;
}

.details {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin: 0.5rem 0 0;
}

.label {
  color: var(--color-grey-medium);
}

.explanation {
  composes: subline from '../../components/Typography/Typography.module.css';
}

.buttonsLine {
  composes: buttonsLine from '../../components/Button/Button.module.css';
}

.cancel {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
}

.cta {
  composes: buttonSecondary from '../../components/Button/Button.module.css';
}
