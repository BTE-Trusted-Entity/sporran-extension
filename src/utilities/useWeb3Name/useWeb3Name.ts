import { ConfigService, Did, DidUri } from '@kiltprotocol/sdk-js';

import { isFullDid } from '../did/did';
import { useAsyncValue } from '../useAsyncValue/useAsyncValue';

export function useWeb3Name(
  did: DidUri | undefined,
): string | null | undefined {
  return useAsyncValue(
    async (did) => {
      if (!(did && isFullDid(did))) {
        return null;
      }

      const api = ConfigService.get('api');
      const result = await api.call.did.query(Did.toChain(did));
      if (result.isNone) {
        return null;
      }

      return Did.linkedInfoFromChain(result).web3Name;
    },
    [did],
  );
}
