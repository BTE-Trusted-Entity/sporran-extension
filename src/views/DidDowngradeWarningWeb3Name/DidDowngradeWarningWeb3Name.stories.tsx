import { Meta } from '@storybook/react';

import { moreIdentitiesMock as identities } from '../../utilities/identities/IdentitiesProvider.mock';

import { DidDowngradeWarningWeb3Name } from './DidDowngradeWarningWeb3Name';

export default {
  title: 'Views/DidDowngradeWarningWeb3Name',
  component: DidDowngradeWarningWeb3Name,
} as Meta;

export function Template(): JSX.Element {
  return (
    <DidDowngradeWarningWeb3Name
      identity={identities['4q11Jce9wqM4A9GPB2z8n4K8LF9w2sQgZKFddhuKXwQ2Qo4q']}
    />
  );
}
