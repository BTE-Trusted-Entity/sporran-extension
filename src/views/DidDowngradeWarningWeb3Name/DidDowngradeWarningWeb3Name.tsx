import { browser } from 'webextension-polyfill-ts';
import { Link, useHistory } from 'react-router-dom';

import * as styles from './DidDowngradeWarningWeb3Name.module.css';

import { Identity } from '../../utilities/identities/types';
import { generatePath, paths } from '../paths';

import { LinkBack } from '../../components/LinkBack/LinkBack';
import { Stats } from '../../components/Stats/Stats';
import { Avatar } from '../../components/Avatar/Avatar';
import { useWeb3Name } from '../../utilities/useWeb3Name/useWeb3Name';

interface Props {
  identity: Identity;
}

export function DidDowngradeWarningWeb3Name({
  identity,
}: Props): JSX.Element | null {
  const t = browser.i18n.getMessage;

  const { goBack } = useHistory();

  const { address, did } = identity;

  const web3name = useWeb3Name(did);

  return (
    <section className={styles.container}>
      <h1 className={styles.heading}>
        {t('view_DidDowngradeWarningWeb3Name_heading')}
      </h1>
      <p className={styles.subline}>
        {t('view_DidDowngradeWarningWeb3Name_subline')}
      </p>

      <Avatar identity={identity} />

      <p className={styles.warning}>
        {t('view_DidDowngradeWarningWeb3Name_warning')}
      </p>

      <p className={styles.details}>
        <span className={styles.label}>
          {t('view_DidDowngradeWarningWeb3Name_web3name')}
        </span>
        {web3name}
      </p>

      <p className={styles.explanation}>
        {t('view_DidDowngradeWarningWeb3Name_explanation')}
      </p>

      <p className={styles.buttonsLine}>
        <button onClick={goBack} className={styles.cancel}>
          {t('common_action_cancel')}
        </button>
        <Link
          to={generatePath(
            paths.identity.did.manage.downgrade.warning.credentials,
            { address },
          )}
          className={styles.cta}
        >
          {t('common_action_continue')}
        </Link>
      </p>

      <LinkBack />
      <Stats />
    </section>
  );
}
