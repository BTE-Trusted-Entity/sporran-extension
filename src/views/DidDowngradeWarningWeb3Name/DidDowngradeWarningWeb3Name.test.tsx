import { identitiesMock as identities, render } from '../../testing/testing';

import { mockIsFullDid } from '../../utilities/did/did.mock';

import { useWeb3Name } from '../../utilities/useWeb3Name/useWeb3Name';

import { DidDowngradeWarningWeb3Name } from './DidDowngradeWarningWeb3Name';

jest.mock('../../utilities/useWeb3Name/useWeb3Name');
jest.mocked(useWeb3Name).mockReturnValue('fancy-name');

describe('DidDowngradeWarningWeb3Name', () => {
  it('should render', async () => {
    mockIsFullDid(true);
    const { container } = render(
      <DidDowngradeWarningWeb3Name
        identity={
          identities['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo']
        }
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
