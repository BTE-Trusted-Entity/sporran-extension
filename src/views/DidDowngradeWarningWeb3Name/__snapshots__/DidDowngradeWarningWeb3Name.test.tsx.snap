// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DidDowngradeWarningWeb3Name should render 1`] = `
<div>
  <section
    class="container"
  >
    <h1
      class="heading"
    >
      Remove DID
    </h1>
    <p
      class="subline"
    >
      Remove DID from the KILT blockchain
    </p>
    <div
      class="avatar fullDid"
    >
      Identicon
    </div>
    <p
      class="warning"
    >
      Attention! Removing your DID will also remove your web3name:
    </p>
    <p
      class="details"
    >
      <span
        class="label"
      >
        web3name: 
      </span>
      fancy-name
    </p>
    <p
      class="explanation"
    >
      Please note: if you remove your web3name, it can be claimed by someone else.
    </p>
    <p
      class="buttonsLine"
    >
      <button
        class="cancel"
      >
        Cancel
      </button>
      <a
        class="cta"
        href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo/did/downgrade/warning/credentials"
      >
        Continue
      </a>
    </p>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </section>
</div>
`;
